jr-test-module/src/test/java/Java17RecordTest.java

```diff
@@ -12,9 +13,15 @@ public class Java17RecordTest {
 
     @Test
     public void testJava14RecordSupport() throws IOException {
+        JSON jsonParser = JSON.builder().enable(Feature.USE_FIELD_MATCHING_GETTERS).build();
         var expectedString = "{\"message\":\"MOO\",\"object\":{\"Foo\":\"Bar\"}}";
-        var json = JSON.builder().enable(JSON.Feature.USE_FIELD_MATCHING_GETTERS).build().asString(new Cow("MOO", Map.of("Foo", "Bar")));
+        Cow expectedObject = new Cow("MOO", Map.of("Foo", "Bar"));
+
+        var json = jsonParser.asString(expectedObject);
         Assert.assertEquals(expectedString, json);
+
+        Cow object = jsonParser.beanFrom(Cow.class, json);
```

Comment:
Here is some feedback on other variants of input data.

a) different order of fields in JSON than in Java record
input: `{"object":{"Foo":"Bar"}, "message":"MOO"}`
result: `Failed to create an instance of jr.Java17RecordTest$Cow due to (java.lang.IllegalArgumentException): argument type mismatch`

This should get fixed.

b) missing field in JSON
input: `{"message":"MOO"}`
result: `Failed to create an instance of jr.Java17RecordTest$Cow due to (java.lang.IllegalArgumentException): wrong number of arguments`

I am wondering how deserialization to records should handle missing data as the all-args record constructor requires all fields and if there should also be support for `Optional` like provided by `jackson-modules-java8`.

Reply:
@darxriggs Could you file a separate issue for the problem with ordering? We definitely need more testing here. It can include part about missing values as well (no need for 2 issues).
As to missing values: should pass `null`, although this will be problematic for many cases like primitives.
These can be addressed incrementally; `Optional` isn't yet supported for anything else yet so it can be deferred.

Reply:
Thanks @darxriggs, I've included your feedback in a follow-up: https://github.com/FasterXML/jackson-jr/pull/163